### PR TITLE
Address some new Roslyn nullability warnings

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
I tested the runtime build against a recent version of the Roslyn compiler (toolset version 3.7.0-1.20251.4) and saw that it's spitting out additional warnings regarding our use of nullability annotations. This PR fixes those warnings in our code base.

Note: I had to change the public-facing annotations on `LazyInitializer.EnsureInitialized<T>(...)`, as they weren't entirely correct. The original annotations had the parameter `[NotNull] ref object? syncLock`, meaning that _syncLock_ would always definitely be assigned upon method exit. This isn't always correct, as assignment may be bypassed if the target is already allocated. Instead, I changed the annotation to `[NotNullIfNotNull("syncLock")] ref object? syncLock`, meaning that if _syncLock_ is non-null on method enter, it will also be non-null on method exit.

Aside from changing those public-facing annotations this PR introduces no behavioral changes.

Contributes to https://github.com/dotnet/runtime/issues/34417.

### CODE REVIEW

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs:26]: I suggest we actually be truthful here, remove the [NotNull], and change the call sites to use `!` where necessary.

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs:28]: This isn't necessary.  You can just change the:
```C#
_ptr = Unsafe.AsPointer(ref s);
```
to be:
```C#
_ptr = Unsafe.AsPointer(ref s!);
```

But, as commented earlier, I think the right solution instead is to remove the [NotNull].

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs:44]: Ditto

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs
Comment (stephentoub) [reply src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs:28]: (Call sites can then use `!` as is appropriate, e.g. `string s = null!` or `ref s!`.)

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:68]:
```suggestion
            Debug.Assert(array != null);
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullabilityHelpers.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullabilityHelpers.cs:21]: Cute, but I suggest we delete this.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs:74]: Does it negatively affect the resulting asm to store the `GetStateMachineBox(ref stateMachine, ref taskField).MoveNextAction` into a local before the try block?  If it doesn't, that'd be my preferred solution.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs:74]: Or, actually, are any call sites relying on the [NotNull] annotation?  Even better would just be to delete that.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncValueTaskMethodBuilderT.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncValueTaskMethodBuilderT.cs:164]: Ditto

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:54]: Why is this necessary?

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:110]: Ditto.  If this is warning, that seems like a Roslyn bug, unless I'm missing something.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:110]: (And assuming it is a Roslyn bug, I'd rather just use `!` to suppress it and add a `// TODO-NULLABLE: https://...` comment linking to the appropriate issue.)

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:147]: Hmm.  This `[NotNullIfNotNull("syncLock")]` on the `syncLock` parameter is basically saying "the method won't set this ref parameter to null".  I wonder if we have other places where we should be adding such an attribute... to my knowledge these if the first time we've used `NotNullIfNotNull` to refer to the same parameter it's attributed on.

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:258]: Same question as earlier; I'm not understanding why this change is necessary.

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:54]: Ah, I see.  We've annotated Volatile.Read's return as being `[NotNullIfNotNull("location")]`, which will tell the compiler that the result is non-null if location is non-null, but the compiler can't infer from that that location is non-null if the result is non-null, and we lack a `[NotNullIfResultNotNull]` attribute.  Ugh.

I'd rather just address this with a `!` on the `ref target`.

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:147]: For example, everywhere we have a `ref T foo` parameter, should we be annotating it as `[NotNullIfNotNull("foo")]`?  That'd be kind of awful.  @jcouv, what's the recommended approach here?

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:290]: We should delete the TODO-NULLABLE comment below.  I'm also a little surprised the `!` on `target` below isn't suppressing the warning on target for [NotNull].

**Review by stephentoub — COMMENTED** [review]
Thanks for working on this.

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:54]: The way we'd address this would be to put a bang on the first _target_, not the second _target_:

```cs
return Volatile.Read(ref target!) ?? EnsureInitializedCore(ref target);
```

That's technically a lie, since the bang is in the wrong spot. But since it's really just an internal implementation detail of this method I suppose that's ok and we shouldn't worry too much about it?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:54]: > I suppose that's ok and we shouldn't worry too much about it?

Yup.  Thanks.

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Array.cs:68]: This won't solve the problem. The compiler expects to see a write to the _array_ parameter on every code path, or it expects the _array_ parameter to be passed to a method whose ref parameter is marked `[NonNull]`. I couldn't figure out any way to solve this other than by using a helper method as introduced here (which gets compiled away entirely due to the `[Conditional]` attribute) or by using a `#pragma warning disable CS8777` at the call site.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Array.cs:68]: > This won't solve the problem.

Are you sure?  I just tried it with last night's SDK and it appears to.  Generally `Debug.Assert` can be used to tell the compiler that something is non-null because the Assert argument is attributed as `DoesNotReturnIf(false)`

File: src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs:139]: Along those lines, `Volatile.Read<T>(ref T value)` could be a candidate for an `[In]`-like attribute. Basically: "The caller must explicitly use the _ref_ keyword to pass an address to this routine (no implicit _in_ conversion allowed), but the routine promises not to assign a new value to the provided address." That attribute would also imply `[NotNullIfNotNull("sameParam")]` by construction.

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Array.cs]: > Derp. I typo good in source.

:smile:

**Review by stephentoub — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
@stephentoub Were you waiting for your question at https://github.com/dotnet/runtime/pull/35782#discussion_r419546314 to be addressed before I merge? Otherwise I'll merge by EOD once CI goes green.

**Comment by stephentoub** [comment]
> Stephen Toub Were you waiting for your question at #35782 (comment) to be addressed before I merge?

No, merge when ready.

I'm a little uneasy about using [NotNullIfNotNull] in this way, but I don't want it to hold up your PR.  @jcouv, I'll be interested in your thoughts here.

**Comment by jcouv** [comment]
Hadn't thought of such usage :-) That's clever and I think it's legal/correct.

**Comment by stephentoub** [comment]
Thanks.  FWIW, I looked through all of our ref parameters, and there are very few additional ones that would benefit from this treatment (e.g. `EventSource.Write<T>(..., ref T data)`, `MemoryMarshal.CreateSpan<T>(ref T reference, int length)`).

**Comment by GrabYourPitchforks** [comment]
@stephentoub - I wonder if those would benefit from the `[In]`-like attribute I pondered up at https://github.com/dotnet/runtime/pull/35782#discussion_r419620297. What these APIs (and `Volatile.Read`) all have in common is that they guarantee they won't reassign the reference target. Whereas `LazyInitializer.EnsureInitialied(..., ref object? syncObject)` says "I may or may not write the reference, but if I do I guarantee I'll write a non-null value to it."

**Comment by stephentoub** [comment]
>  I wonder if those would benefit from the [In]-like attribute I pondered

How is that different from `in` / `ref readonly`?

**Comment by GrabYourPitchforks** [comment]
"in" means that the call site will neither overwrite the target address _nor_ call any API on the target reference which will cause mutation of the value. "in" also doesn't require you to use the "in" keyword when calling the target method, which runs the risk of making hidden copies of locals at the call site.

What I'm proposing is more akin to what ECMA refers to as a controlled mutability reference. You still need to use the "ref" keyword at the call site, but the callee guarantees that the target address will not be overwritten. It may however be mutated through other means.

The upshot is that such APIs implicitly have a contract "if you call me with a non null ref, I guarantee the ref will remain non null at method exit."

**Comment by stephentoub** [comment]
> The upshot is that such APIs implicitly have a contract "if you call me with a non null ref, I guarantee the ref will remain non null at method exit."

That's what `[NotNullIfNotNull("p")] p` provides.  We don't need another attribute to express that.

What you're discussing is a "I guarantee I won't change this reference", and I think the distinctions between that and `in` / `ref readonly` are too subtle to be valuable.  I certainly wouldn't want to introduce such a thing just for nullable reference types.  The couple of examples I cited are a rarity and even then there's limited benefit to add `[NotNullIfNotNull("p")]`, nevermind any new attribute.  

(And as an aside, `[In]` already exists.)

**Comment by GrabYourPitchforks** [comment]
> I certainly wouldn't want to introduce such a thing just for nullable reference types.

It would also be the natural return type of `Unsafe.Unbox<T>(...)` and would make that method 100% verifiably type-safe. :)
But yeah, probably too esoteric a feature to be very useful.

**Pull request merged**
